warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/TruncatingTimestampIterator.java

```diff
@@ -0,0 +1,118 @@
+    private final SortedKeyValueIterator<Key,Value> delegate;
+
+    public TruncatingTimestampIterator(SortedKeyValueIterator<Key,Value> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        delegate.init(source, options, env);
+    }
+
+    @Override
+    public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+        return new TruncatingTimestampIterator(delegate.deepCopy(env));
+    }
+
+    @Override
+    public Key getTopKey() {
+        return getTruncatedTimestampKey(delegate.getTopKey());
+    }
+
+    @Override
+    public Value getTopValue() {
+        return delegate.getTopValue();
+    }
+
+    @Override
+    public boolean hasTop() {
+        return delegate.hasTop();
+    }
+
+    @Override
+    public void next() throws IOException {
+        delegate.next();
+    }
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+        delegate.seek(range, columnFamilies, inclusive);
+    }
+
+    /**
+     * Return the key with a truncated timestamp
+     *
+     * @param key
+     *            The key
+     * @return The key with a truncated timestamp
+     */
+    public static Key getTruncatedTimestampKey(Key key) {
+        if (key == null) {
+            return key;
+        }
+
+        long timestamp = key.getTimestamp();
+        long newTimestamp = truncateTimestamp(timestamp);
+        if (timestamp == newTimestamp) {
+            return key;
+        }
+
+        return new Key(key.getRow().getBytes(), key.getColumnFamily().getBytes(), key.getColumnQualifier().getBytes(), key.getColumnVisibility().getBytes(),
+                        newTimestamp, key.isDeleted(), false);
+    }
+
+    /**
+     * truncate a timestamp (possibly composite timestamp) to the beginning of the day.
+     *
+     * @param timestamp
+     *            timestamp (possibly composite timestamp)
+     * @return timestamp with the event date truncated to the beginning of the day
+     */
+    public static long truncateTimestamp(long timestamp) {
+        if (CompositeTimestamp.isCompositeTimestamp(timestamp)) {
+            long eventDate = CompositeTimestamp.getEventDate(timestamp);
+            int ageOffDeltaDays = CompositeTimestamp.getAgeOffDeltaDays(timestamp);
+            return CompositeTimestamp.getCompositeDeltaTimeStamp(beginningOfDay(eventDate), ageOffDeltaDays);
+        } else {
+            return beginningOfDay(timestamp);
+        }
+    }
+
+    /**
+     * Return the beginning of the day for a timestamp
+     *
+     * @param timestamp
+     *            timestamp
+     * @return beginning of the day
+     */
+    public static long beginningOfDay(long timestamp) {
+        Calendar c = calendar.get();
```

Comment:
We could use java8 date/time to do this truncation which would eliminate the need for the ThreadLocal and I believe is a performance win from some quick searching online.

```
Instant truncated = Instant.ofEpochMilli(timestamp).truncatedTo(ChronoUnit.DAYS);
return truncated.toEpochMilli();
```
```

Reply:
done